server/auth/auth.go

```diff
@@ -940,18 +1080,68 @@ func (auth *AuthManager) handleConnect(conn comms.Link, msg *msgjson.Message) *m
 		recentOrders: latestFinished,
 		suspended:    !open,
 	}
+
+	// Compute the user's cancellation rate.
 	cancels, completions, rate, penalize := auth.checkCancelRate(client)
-	if penalize && open && !auth.anarchy {
+	if penalize && open && !auth.freeCancels {
 		// Account should already be closed, but perhaps the server crashed
 		// or the account was not penalized before shutdown.
 		client.suspended = true
 		// The account might now be closed if the cancellation rate was
 		// exceeded while the server was running in anarchy mode.
 		auth.storage.CloseAccount(acctInfo.ID, account.CancellationRate)
 		log.Debugf("Suspended account %v (cancellation rate = %.2f%%, %d cancels : %d successes) connected.",
-			acctInfo.ID, cancels, completions, 100*rate)
+			acctInfo.ID, 100*rate, cancels, completions)
+	}
+
+	// Compute the user's ban score.
+	score, err := auth.loadUserScore(user)
+	if err != nil {
+		log.Errorf("failed to compute user %v score: %v", user, err)
+		return &msgjson.Error{
+			Code:    msgjson.RPCInternalError,
+			Message: "DB error",
+		}
+	}
+	log.Debugf("User %v connected with ban score %d", user, score)
+	if score >= int32(auth.banScore) && open && !auth.anarchy {
+		// Would already be penalized unless server changed the banScore or
+		// was previously running in anarchy mode.
+		client.suspended = true
+		auth.storage.CloseAccount(user, account.FailureToAct)
+		log.Debugf("Suspended account %v (score = %d) connected.", acctInfo.ID, score)
+	}
```

Comment:
Ah, a problem that this creates, and which the cancellation rate banning already created, is that we can't really unban.  This would seem to be the crucial need for the penalties table work in PR #572 .

@JoeGruffins maybe we can consider what we want to record new tables (penalties, violations, or both) given the other elements of this PR, namely making a distinction between recording violations and creating penalties https://github.com/decred/dcrdex/pull/700/files#r493726170.  I see a need to forgive individual violations, as well as a separate ability to lift an active penalty when forgiving a violation, but lifting the actual penalty could just be done implicitly by re-integrading the unforgiven penalties. But I think we should not have `duration` as a property of a violation though.  `duration` makes sense for a penalty, but not a violation.

Reply:
On possible solution to supporting violation forgiveness might be to decrement from a category the `violationCounter`, such that the user would only need DB support for forgiven violation counts (e.g. 1 forgiven NoSwapAsMaker, 2 forgiven NoRedeemAsTaker, etc) that adjusts the violation score on connect... as an alternative to forgiving *specific* violation instances.

Reply:
https://github.com/decred/dcrdex/pull/700/commits/e39a520122d406cad18c36fd79f083d8def23be9 tries this out.